### PR TITLE
fix(gateway): invalidate shared-token/password WS sessions on secret rotation [AI]

### PR DESCRIPTION
**Pull request by pgondhi987**
## Summary

- **Problem:** Existing WebSocket sessions authenticated with a shared gateway token or password continued operating normally after the secret was rotated, because no invalidation path existed for shared-secret sessions (only device-token sessions were proactively disconnected).
- **Why it matters:** Emergency credential rotation lost most of its value — an attacker holding an established session retained full access to control-plane methods until the gateway restarted.
- **What changed:** Sessions authenticated via `token` or `password` mode now carry a non-reversible SHA-256 generation ID derived from the active secret. On every config reload or restart that changes the secret, stale sessions are proactively closed with code 4001. A per-message guard provides defense-in-depth for any session that slips through the proactive sweep.
- **What did NOT change:** Device-token session lifecycle, all non-shared-auth connection types, and the public Plugin SDK surface are unaffected.

## Change Type (select all)

- [x] Bug fix
- [x] Security hardening

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [x] Auth / tokens

## Linked Issue/PR

- [ ] This PR fixes a bug or regression

## Root Cause (if applicable)

- Root cause: `GatewayWsClient` had no field representing the auth-generation at connect time for shared-secret modes; the config-reload and config-restart handlers discarded the result of `activateRuntimeSecrets` without comparing generations.
- Missing detection / guardrail: No integration test exercised token rotation with `reload.mode = off` or with a hot-reload SecretRef change.
- Contributing context (if known): Device-token revocation already had an explicit invalidation path; the equivalent path for shared secrets was never implemented.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Seam / integration test
- Target test or file: `src/gateway/server.shared-token-session-rotation.test.ts`, `src/gateway/server.shared-token-hot-reload.test.ts`
- Scenario the test should lock in: (1) `config.set` with `reload.mode = off` rotates the token → existing session receives close 4001 before any further request completes. (2) Hot-reload via file write with a rotated SecretRef env var → existing session receives close 4001.
- Why this is the smallest reliable guardrail: Both tests use a real in-process gateway server over localhost WebSocket, exercising the full connect → rotate → close lifecycle without mocking auth internals.
- Existing test that already covers this (if any): None prior to this PR.

## User-visible / Behavior Changes

Existing WebSocket sessions authenticated with a shared token or password will be disconnected with close code 4001 ("gateway auth changed") when the shared secret changes via config rotation. Previously those sessions remained open until gateway restart.

## Diagram (if applicable)

```text
Before:
[secret rotated] -> [existing sessions continue operating with old credentials]

After:
[secret rotated] -> [sessions stamped with old generation] -> [close 4001 sent proactively]
                 -> [per-message guard catches any that slip through]
```

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? Yes
- New/changed network calls? No
- Command/tool execution surface changed? No
- Data access scope changed? No
- Secrets handling change: The active shared secret is hashed (SHA-256, base64url, with a mode-separator null byte) into a generation ID stored on each connection. The raw secret is never stored or transmitted; the hash is only used for equality comparison. Rotation of the secret produces a new hash, causing the mismatch check to fire.

### Environment

- OS: Linux (CI / gateway test suite)
- Runtime/container: Node 22, Bun
- Model/provider: N/A
- Integration/channel (if any): N/A
- Relevant config (redacted): `gateway.auth.mode = token`, `gateway.reload.mode = off`

### Steps

1. Establish a WebSocket session using the current shared gateway token.
2. Call `config.set` with a rotated token value and `reload.mode = off`.
3. Attempt any subsequent request on the established session.

### Expected

- Session closes with code 4001 and reason "gateway auth changed" before the next request is processed.

### Actual (before fix)

- Session remains open and all subsequent requests succeed using the old token credential.

## Evidence

- [x] Failing test/log before + passing after — `server.shared-token-session-rotation.test.ts` and `server.shared-token-hot-reload.test.ts` added; both would have failed against the pre-fix code.

## Human Verification (required)

> This PR was generated by OpenAI Codex and reviewed by Claude (AI-assisted). No human manual verification was performed beyond automated test and review coverage.

- Verified scenarios: Token rotation with `reload.mode = off` (integration test); hot-reload SecretRef rotation (integration test); per-message guard with `requiredSharedGatewaySessionGeneration` sentinel during async secrets-resolution window (code review).
- Edge cases checked: `activate: false` restart path async window; SecretRef optimistic vs. authoritative generation divergence; error-path restoration of `requiredSharedGatewaySessionGeneration`; `null`/`undefined`/`string` three-state sentinel semantics.
- What you did **not** verify: Live gateway with real vault-backed SecretRef; concurrent `onRestart` calls under load; non-Linux platforms.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? Yes
- Config/env changes? No
- Migration needed? No

## Risks and Mitigations

- Risk: Sessions are proactively terminated on secret rotation, which is a behavior change for clients that expect to survive a rotation window.
  - Mitigation: This is the intended security behavior; clients must reconnect with the new credential. The close code 4001 is distinct and allows clients to handle it explicitly.
- Risk: Optimistic generation computed from raw `nextConfig` (before SecretRef resolution) may differ from the authoritative post-resolution generation, causing a brief window where per-message checks use the optimistic value.
  - Mitigation: The authoritative value is computed and applied immediately after `activateRuntimeSecrets` completes, with a second disconnect sweep. On error, the sentinel is restored to its prior value.

### CODE REVIEW

File: src/gateway/server/ws-connection/message-handler.ts [hunk 1]
```diff
@@ -1330,6 +1338,21 @@ export function attachGatewayWsMessageHandler(params: {
...
+          requiredSharedGatewaySessionGeneration === null
+            ? resolveSharedGatewaySessionGeneration(getResolvedAuth())
+            : requiredSharedGatewaySessionGeneration;
+        if (client.sharedGatewaySessionGeneration !== activeSharedGatewaySessionGeneration) {
+          setCloseCause("gateway-auth-rotated", {
+            authGenerationStale: true,
+          });
+          close(4001, "gateway auth changed");
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server/ws-connection/message-handler.ts:1360]: Addressed in cb8a435f07bba76f4f910405e2f641f947d36234.

File: src/gateway/server.impl.ts [hunk 2]
```diff
@@ -1512,15 +1573,48 @@ export async function startGatewayServer(
...
+                requiredSharedGatewaySessionGeneration;
+              const previousSharedGatewaySessionGeneration = currentSharedGatewaySessionGeneration;
+              const optimisticSharedGatewaySessionGeneration =
+                resolveSharedGatewaySessionGenerationForConfig(nextConfig);
+              if (
+                previousSharedGatewaySessionGeneration !== optimisticSharedGatewaySessionGeneration
+              ) {
+                requiredSharedGatewaySessionGeneration = optimisticSharedGatewaySessionGeneration;
+                disconnectStaleSharedGatewayAuthClients(optimisticSharedGatewaySessionGeneration);
+              }
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in 95bf4fbe7698fd961cb7134b6595721faf7e63eb.

File: src/gateway/server.impl.ts [hunk 3]
```diff
@@ -793,6 +811,36 @@ export async function startGatewayServer(
...
+      requiredSharedGatewaySessionGeneration = null;
+    }
+  };
+  const enforceSharedGatewaySessionGenerationForConfigWrite = (nextConfig: OpenClawConfig) => {
+    if (resolveGatewayReloadSettings(nextConfig).mode !== "off") {
+      return;
+    }
+    const nextSharedGatewaySessionGeneration =
+      resolveSharedGatewaySessionGenerationForConfig(nextConfig);
+    requiredSharedGatewaySessionGeneration = nextSharedGatewaySessionGeneration;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in 5f213dd132ca1a0b3bedb1c8a5dedd7c633afd6c.

File: src/gateway/server.impl.ts [hunk 4]
```diff
@@ -1512,15 +1573,42 @@ export async function startGatewayServer(
...
+                const nextSharedGatewaySessionGeneration =
+                  resolveSharedGatewaySessionGenerationForConfig(prepared.config);
+                if (previousSharedGatewaySessionGeneration !== nextSharedGatewaySessionGeneration) {
+                  requiredSharedGatewaySessionGeneration = nextSharedGatewaySessionGeneration;
+                  disconnectStaleSharedGatewayAuthClients(nextSharedGatewaySessionGeneration);
+                } else {
+                  requiredSharedGatewaySessionGeneration =
+                    previousRequiredSharedGatewaySessionGeneration;
+                }
+                requestGatewayRestart(plan, nextConfig);
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in 5f213dd132ca1a0b3bedb1c8a5dedd7c633afd6c.

File: src/gateway/server.impl.ts [hunk 5]
```diff
@@ -1512,15 +1582,47 @@ export async function startGatewayServer(
...
+                const prepared = await activateRuntimeSecrets(nextConfig, {
+                  reason: "restart-check",
+                  activate: false,
+                });
+                const nextSharedGatewaySessionGeneration =
+                  resolveSharedGatewaySessionGenerationForConfig(prepared.config);
+                const restartQueued = requestGatewayRestart(plan, nextConfig);
+                if (!restartQueued) {
+                  requiredSharedGatewaySessionGeneration =
+                    previousRequiredSharedGatewaySessionGeneration;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in cad1ecbcf647604529671a751eca017afccf0084.

File: src/gateway/server.impl.ts [hunk 6]
```diff
@@ -1512,15 +1582,49 @@ export async function startGatewayServer(
...
+                  requiredSharedGatewaySessionGeneration = runtimeSharedGatewaySessionGeneration;
+                  disconnectStaleSharedGatewayAuthClients(runtimeSharedGatewaySessionGeneration);
+                  return;
+                }
+                if (previousSharedGatewaySessionGeneration !== nextSharedGatewaySessionGeneration) {
+                  requiredSharedGatewaySessionGeneration = nextSharedGatewaySessionGeneration;
+                  disconnectStaleSharedGatewayAuthClients(nextSharedGatewaySessionGeneration);
+                } else {
+                  requiredSharedGatewaySessionGeneration =
+                    previousRequiredSharedGatewaySessionGeneration;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in eaf4eddbe939dde4f2ecd1c4dadde600728de91d.

File: src/gateway/server/ws-connection/message-handler.ts [hunk 7]
```diff
@@ -1330,6 +1338,17 @@ export function attachGatewayWsMessageHandler(params: {
         return;
       }
 
+      if (client.usesSharedGatewayAuth) {
+        const requiredSharedGatewaySessionGeneration = getRequiredSharedGatewaySessionGeneration();
+        if (client.sharedGatewaySessionGeneration !== requiredSharedGatewaySessionGeneration) {
+          setCloseCause("gateway-auth-rotated", {
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server/ws-connection/message-handler.ts:1355]: Addressed in 1651e50fbec887dd4c6513c7929a298ae657adaa.

File: src/gateway/server-methods/config.ts [hunk 8]
```diff
@@ -417,6 +430,10 @@ export const configHandlers: GatewayRequestHandlers = {
     if (!(await ensureResolvableSecretRefsOrRespond({ config: parsed.config, respond }))) {
       return;
     }
+    const refreshSharedGatewayAuthGeneration = didSharedGatewayAuthChange(
+      snapshot.config,
+      parsed.config,
+    );
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server-methods/config.ts]: Addressed in 1651e50fbec887dd4c6513c7929a298ae657adaa.

File: src/gateway/server/ws-connection/message-handler.ts [hunk 9]
```diff
@@ -1183,6 +1186,21 @@ export function attachGatewayWsMessageHandler(params: {
...
+            : undefined;
+        if (authMethod === "token" || authMethod === "password") {
+          const requiredSharedGatewaySessionGeneration =
+            getRequiredSharedGatewaySessionGeneration();
+          if (sharedGatewaySessionGeneration !== requiredSharedGatewaySessionGeneration) {
+            setCloseCause("gateway-auth-rotated", {
+              authGenerationStale: true,
+            });
+            close(4001, "gateway auth changed");
+            return;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server/ws-connection/message-handler.ts]: Addressed in c293368851f58669202d2e875cd5013c09836ddb.

File: src/gateway/server.impl.ts [hunk 10]
```diff
@@ -1499,6 +1568,7 @@ export async function startGatewayServer(
             readSnapshot: readConfigFileSnapshot,
             subscribeToWrites: registerConfigWriteListener,
             onHotReload: async (plan, nextConfig) => {
+              const previousSharedGatewaySessionGeneration = currentSharedGatewaySessionGeneration;
               const previousSnapshot = getActiveSecretsRuntimeSnapshot();
               const prepared = await activateRuntimeSecrets(nextConfig, {
                 reason: "reload",
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts:1584]: Addressed in aa7460871ae560ed57854de4dabf405ed279a473.

File: src/gateway/server.impl.ts [hunk 11]
```diff
@@ -1512,15 +1588,46 @@ export async function startGatewayServer(
...
+                  activate: false,
+                });
+                const nextSharedGatewaySessionGeneration =
+                  resolveSharedGatewaySessionGenerationForConfig(prepared.config);
+                const restartQueued = requestGatewayRestart(plan, nextConfig);
+                if (!restartQueued) {
+                  const runtimeSharedGatewaySessionGeneration =
+                    resolveSharedGatewaySessionGenerationForRuntimeSnapshot();
+                  requiredSharedGatewaySessionGeneration = runtimeSharedGatewaySessionGeneration;
+                  disconnectStaleSharedGatewayAuthClients(runtimeSharedGatewaySessionGeneration);
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in 3890418ea74e090d5bf80834c5f90c2556252b1f.

File: src/gateway/server.impl.ts [hunk 12]
```diff
@@ -1512,15 +1588,50 @@ export async function startGatewayServer(
...
+                  activate: false,
+                });
+                const nextSharedGatewaySessionGeneration =
+                  resolveSharedGatewaySessionGenerationForConfig(prepared.config);
+                const restartQueued = requestGatewayRestart(plan, nextConfig);
+                if (!restartQueued) {
+                  if (
+                    previousSharedGatewaySessionGeneration !== nextSharedGatewaySessionGeneration
+                  ) {
+                    requiredSharedGatewaySessionGeneration = nextSharedGatewaySessionGeneration;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in 2404b130b01abefe6817c7fe350bb41fb50083d6.

File: src/gateway/server.impl.ts [hunk 13]
```diff
@@ -1512,15 +1588,50 @@ export async function startGatewayServer(
                 } else {
                   clearSecretsRuntimeSnapshot();
                 }
+                currentSharedGatewaySessionGeneration = previousSharedGatewaySessionGeneration;
                 throw err;
               }
+              setCurrentSharedGatewaySessionGeneration(nextSharedGatewaySessionGeneration);
+              if (previousSharedGatewaySessionGeneration !== nextSharedGatewaySessionGeneration) {
+                disconnectStaleSharedGatewayAuthClients(nextSharedGatewaySessionGeneration);
+              }
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in 2404b130b01abefe6817c7fe350bb41fb50083d6.

File: src/gateway/server.impl.ts [hunk 14]
```diff
@@ -793,6 +820,36 @@ export async function startGatewayServer(
...
+  const setCurrentSharedGatewaySessionGeneration = (nextGeneration: string | undefined) => {
+    currentSharedGatewaySessionGeneration = nextGeneration;
+    if (requiredSharedGatewaySessionGeneration === nextGeneration) {
+      requiredSharedGatewaySessionGeneration = null;
+    }
+  };
+  const enforceSharedGatewaySessionGenerationForConfigWrite = (nextConfig: OpenClawConfig) => {
+    if (resolveGatewayReloadSettings(nextConfig).mode !== "off") {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts]: Addressed in 9f25e6b6584a33f5eadbd9a2a29ee39048c2e5e8.

File: src/gateway/server-methods/config.ts [hunk 15]
```diff
@@ -578,6 +592,7 @@ export const configHandlers: GatewayRequestHandlers = {
       },
       undefined,
     );
+    queueSharedGatewayAuthGenerationRefresh(disconnectSharedAuthClients, validated.config, context);
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server-methods/config.ts]: Addressed in ac03453abb2a1fe5fa51be4ef76da2c40ae91d2d.

File: src/gateway/server.impl.ts [hunk 16]
```diff
@@ -793,6 +820,41 @@ export async function startGatewayServer(
...
+        gatewayClient.socket.close(4001, "gateway auth changed");
+      } catch {
+        /* ignore */
+      }
+    }
+  };
+  const setCurrentSharedGatewaySessionGeneration = (nextGeneration: string | undefined) => {
+    currentSharedGatewaySessionGeneration = nextGeneration;
+    if (requiredSharedGatewaySessionGeneration === nextGeneration) {
+      requiredSharedGatewaySessionGeneration = null;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts:842]: Addressed in e950e376af530eeadf895838c4803d683dff0948.

File: src/gateway/server.shared-token-hot-reload.test.ts [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+
+async function loadCurrentConfig(ws: WebSocket): Promise<Record<string, unknown>> {
+  const current = await rpcReq<{
+    config?: Record<string, unknown>;
+  }>(ws, "config.get", {});
+  expect(current.ok).toBe(true);
+  return structuredClone(current.payload?.config ?? {});
+}
+
+beforeAll(async () => {
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.shared-token-hot-reload.test.ts:67]: Addressed in 530425a2729a2c3d80b3bad781ebcaf8d1fb5304.

File: src/gateway/server.shared-token-session-rotation.test.ts [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+});
+
+afterAll(async () => {
+  testState.gatewayAuth = ORIGINAL_GATEWAY_AUTH;
+  await server.close();
+});
+
+beforeEach(() => {
+  testState.gatewayAuth = { mode: "token", token: OLD_TOKEN };
+});
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.shared-token-session-rotation.test.ts]: Addressed in 530425a2729a2c3d80b3bad781ebcaf8d1fb5304.

File: src/gateway/server.impl.ts [hunk 19]
```diff
@@ -1512,15 +1605,51 @@ export async function startGatewayServer(
                 } else {
                   clearSecretsRuntimeSnapshot();
                 }
+                currentSharedGatewaySessionGeneration = previousSharedGatewaySessionGeneration;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply src/gateway/server.impl.ts:1608]: Addressed in 3867e936dbb79689a4122deea585d6a794fa5bad.
